central/xlsx/deviceprofile.go

```diff
@@ -0,0 +1,223 @@
+	rows, err = dpXlsx.xlsFile.GetRows(deviceResourceSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s worksheet after inserting misshing columns: %w", deviceResourceSheetName, err)
+	}
+
+	// parse the device resource data rows
+	for rowIndex, row := range rows {
+		if rowIndex == 0 {
+			continue
+		}
+
+		convertedDR := dtos.DeviceResource{}
+		_, err = readStruct(&convertedDR, header, row, dpXlsx.fieldMappings)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal an xlsx row into DeviceResource DTO: %w", err)
+		}
+
+		// validate the DeviceResource DTO
+		err = convertedDR.Validate()
+		if err != nil {
+			dpXlsx.validateErrors[convertedDR.Name] = err
+		} else {
+			convertedProfile.DeviceResources = append(convertedProfile.DeviceResources, convertedDR)
+		}
+	}
+	return nil
+}
+
+func (dpXlsx *deviceProfileXlsx) parseDeviceResourceHeader(header *[]string, rowCount int) error {
+	var err error
+	// get the column count of the header row to see if any Object field from MappingTable sheet is not defined
+	colCount := len(*header)
+
+	for objectField, mapping := range dpXlsx.fieldMappings {
+		// check if the mapping object is defined in the DeviceResource sheet if the defaultValue is not empty
+		// if not, insert the mapping object as a new column in the DeviceResource sheet with defaultValue set in each data row
+		if mapping.defaultValue != "" {
+			err = checkMappingObject(dpXlsx.xlsFile, deviceResourceSheetName, &colCount, rowCount, mapping.defaultValue, objectField, header)
+			if err != nil {
+				return fmt.Errorf("failed to check mapping object: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// convertDeviceCommands parses the DeviceCommand sheet and convert the rows to DeviceCommand DTOs
+func (dpXlsx *deviceProfileXlsx) convertDeviceCommands(convertedProfile *dtos.DeviceProfile) error {
+	var header []string
+	rows, err := dpXlsx.xlsFile.GetRows(deviceCommandSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s worksheet: %w", deviceCommandSheetName, err)
+	}
+
+	// checks at least 2 rows exists in the DeviceCommand sheet (1 header and 1 data row)
+	if len(rows) >= 2 {
+		header = rows[0]
+	} else {
+		return fmt.Errorf("at least 2 rows need to be defined in %s worksheet", deviceCommandSheetName)
+	}
+
+	// parse the Device Command data rows
+	for rowIndex, row := range rows {
+		if rowIndex == 0 {
+			continue
+		}
+
+		convertedDC := dtos.DeviceCommand{}
+		_, err = readStruct(&convertedDC, header, row, nil)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal an xlsx row into DeviceCommand DTO: %w", err)
+		}
+
+		// validate the DeviceCommand DTO
+		err = common.Validate(convertedDC)
+		if err != nil {
+			dpXlsx.validateErrors[convertedDC.Name] = err
+		} else {
+			convertedProfile.DeviceCommands = append(convertedProfile.DeviceCommands, convertedDC)
+		}
+	}
+	return nil
+}
+
+func (dpXlsx *deviceProfileXlsx) GetDTOs() any {
```

Comment:
try to use generic type
https://go.dev/blog/intro-generics

Reply:
Since we need to make `instantiation` to the Converter interface, the transform.go file needs to be modified into 2 methods for device and device profile DTOs.